### PR TITLE
Migrate test framework to vitest

### PR DESCRIPTION
**Pull request by timotheeguerin**
Get rid of mocha and upgrade to vitest which is a more modern alternative providing, watch, direct typescript compilation out of the box, expect library and more.

Advantage over mocha:
- Much better cli
  -  watch mode
  - better diff
- Better extension:
  -  tree organization for files too (not everything flattened)
  - update in real time the test(no more need to refresh manually to discover where are the tests)
  - just a little buggy
- Compiles typescript directly
- provides more expectation apis(like jest)

Cons over mocha: 
- Slower(about 2x) but that means we don't need to build the test as part of build which would speed up that part(not as much as is lost)

Todo: 
- typespec-azure migration

### CODE REVIEW

File: packages/bundler/package.json [hunk 1]
```diff
@@ -60,6 +61,7 @@
     "@types/node": "~18.11.9",
     "@typespec/eslint-config-typespec": "workspace:~0.51.0",
     "eslint": "^8.55.0",
+    "vitest": "^1.1.0",
     "mocha": "~10.2.0",
```
Comment (timotheeguerin) [inline packages/bundler/package.json:65]: remove mocha

File: packages/bundler/package.json [hunk 2]
```diff
@@ -60,6 +61,7 @@
...
     "mocha-junit-reporter": "~2.2.1",
     "mocha-multi-reporters": "~1.5.1",
```
Comment (timotheeguerin) [inline packages/bundler/package.json:67]: verify coverage

File: packages/best-practices/package.json [hunk 3]
```diff
@@ -28,8 +28,9 @@
     "clean": "rimraf ./dist ./temp",
     "build": "tsc -p .",
     "watch": "tsc -p . --watch",
-    "test": "mocha",
-    "test-official": "c8 mocha --forbid-only --reporter mocha-multi-reporters",
+    "test": "vitest run",
+    "test:watch": "vitest -w",
+    "test-official": "vitest run --coverage --reporter=junit --reporter=default",
```
Comment (timotheeguerin) [inline packages/best-practices/package.json:33]: Check `.only` is still banned

**Review by bterlson — APPROVED** [review]

File: packages/best-practices/package.json (hunk 3 above)
Comment (timotheeguerin) [reply packages/best-practices/package.json:33]: It is via linter rule

**Pull request merged**